devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+```
+
+## Exposed API
+
+```c
+typedef struct CANCELLATION_TOKEN_TAG CANCELLATION_TOKEN;
+THANDLE_TYPE_DECLARE(CANCELLATION_TOKEN);
+
+typedef struct CANCELLATION_TOKEN_REGISTRATION_TAG CANCELLATION_TOKEN_REGISTRATION;
+THANDLE_TYPE_DECLARE(CANCELLATION_TOKEN_REGISTRATION);
+
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN), cancellation_token_create, bool, canceled);
+MOCKABLE_FUNCTION(, int, cancellation_token_is_canceled, THANDLE(CANCELLATION_TOKEN), cancellation_token, bool*, canceled);
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context);
+MOCKABLE_FUNCTION(, int, cancellation_token_cancel, THANDLE(CANCELLATION_TOKEN), cancellation_token);
+```
+
+## cancellation_token_create
+
+```c
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN), cancellation_token_create, bool, canceled);
+```
+
+`cancellation_token_create` creates a new cancellation token with the specified initial cancellation state.
+
+**SRS_CANCELLATION_TOKEN_04_001: [** `cancellation_token_create` shall allocate memory for a `THANDLE(CANCELLATION_TOKEN)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_002: [** If any underlying error occurs `cancellation_token_create` shall fail and return `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_003: [** `cancellation_token_create` shall create a `TCALL_DISPATCHER` handle by calling `TCALL_DISPATCHER_CREATE(CANCELLATION_TOKEN_CANCEL_CALL)`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_004: [** `cancellation_token_create` shall set the initial state to be equal to `canceled` parameter. **]**
+
+**SRS_CANCELLATION_TOKEN_04_005: [** `cancellation_token_create` shall return a valid `THANDLE(CANCELLATION_TOKEN)` when successful. **]**
+
+## cancellation_token_is_canceled
+
+```c
+MOCKABLE_FUNCTION(, int, cancellation_token_is_canceled, THANDLE(CANCELLATION_TOKEN), cancellation_token, bool*, canceled);
+```
+
+`cancellation_token_is_canceled` assigns `true` to `*canceled` if the cancellation token has been canceled and `false` otherwise.
+
+**SRS_CANCELLATION_TOKEN_04_006: [** `cancellation_token_is_canceled` shall return a non-zero value if `cancellation_token` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_007: [** `cancellation_token_is_canceled` shall return a non-zero value if `canceled` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_008: [** `cancellation_token_is_canceled` shall assign `true` to `*canceled` if the token has been canceled and `false` otherwise. **]**
+
+**SRS_CANCELLATION_TOKEN_04_009: [** `cancellation_token_is_canceled` shall return `0` when successful. **]**
+
+## cancellation_token_register_notify
+
+```c
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context);
```

Comment:
>CANCELLATION_TOKEN_REGISTRATION [](http://example.com/codeflow?start=28&length=31)

How do I unregister? Just by the reference count going to 0? If so maybe the spec for the "dispose" function on that thandle is missing here to say it does that.

Reply:
Good catch. Yes, it is by having the ref count fall to `0` that one un-registers the callback. Added the requirements for both `cancellation_token_registration_dispose` and `cancellation_token_dispose` and tests.